sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc

```diff
@@ -0,0 +1,213 @@
+
+- sycl_ext_oneapi_properties
+- link:sycl_ext_oneapi_kernel_properties.asciidoc[sycl_ext_oneapi_kernel_properties]
+
+== Feature Test Macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification section 6.3.3 "Feature test macros".  Therefore, an
+implementation supporting this extension must predefine the macro
+`SYCL_EXT_INTEL_FPGA_KERNEL_INTERFACE_PROPERTIES` to one of the values defined
+in the table below.  Applications can test for the existence of this macro to
+determine if the implementation supports this feature, or applications can test
+the macro's value to determine which of the extension's APIs the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value |Description
+|1     |Initial extension version.  Base features are supported.
+|===
+
+== Overview
+
+This extension introduces kernel properties to specify how or when control and
+data signals can be passed in to or out of an FPGA kernel. These properties are
+meaningless on non-FPGA devices and can be ignored on such devices.
+
+== Kernel Properties
+
+```c++
+namespace sycl {
+namespace ext {
+namespace oneapi {
+namespace experimental {
+
+// Corresponds to streaming_kernel
+struct streaming_kernel_key {
+  using value_t = property_value<streaming_kernel_key>;
+}; // streaming_kernel_key
+
+// Corresponds to csr_kernel
+struct csr_kernel_key {
+  using value_t = property_value<csr_kernel_key>;
+}; // csr_kernel_key
+
+// Corresponds to stall_free_return_kernel
+struct stall_free_return_kernel_key {
+  using value_t = property_value<stall_free_return_kernel_key>;
+}; // stall_free_return_kernel_key
+
+// Corresponds to pipelined_kernel
+struct pipelined_kernel_key {
+  using value_t = property_value<pipelined_kernel_key>;
+}; // pipelined_kernel_key
+
+inline constexpr streaming_kernel_key::value_t streaming_kernel;
+
+inline constexpr csr_kernel_key::value_t csr_kernel;
+
+inline constexpr stall_free_return_kernel_key::value_t stall_free_return_kernel;
+
+inline constexpr pipelined_kernel_key::value_t pipelined_kernel;
+
+} // namespace experimental
+} // namespace oneapi
+} // namespace ext
+} // namespace sycl
+```
+
+|===
+|Property|Description
+|`streaming_kernel`
+|The `streaming_kernel` property adds the requirement that the kernel must
+ have dedicated ports for input / output signals. This applies for both
+ control, and kernel argument data signals.
+
+|`csr_kernel`
+|The `csr_kernel` property adds the requirement that the kernel must have
+ its input / output control and kernel argument data signals placed in a shared
+ Control and Status Register (CSR) map.
+
+|`stall_free_return_kernel`
+|The `stall_free_return_kernel` property provides the compiler the guarantee
+ that downstream logic will not back-pressure the kernel.
+
+|`pipelined_kernel`
+| The `pipelined_kernel` property directs the compiler to pipeline calls to the
+ kernel so that multiple invocations of the kernel can be in flight
+ simultaneously.
```

Comment:
Do you mean several instances of a `single_task` or this is pipelining at the work-item level?

Reply:
This refers to several instances of a kernel, most commonly a single_task kernel.  It does not refer to pipelining of work-items.  However, there's no reason this couldn't be applied to a parallel_for kernel to request that successive launches of that kernel be pipelined _in addition to_ the usual work item pipelining.  In such a kernel, the work items could be injected into the datapath one after another and, after the last work item from a given kernel invocation has been injected, the first work-item of the next invocation could be injected without waiting for the pipeline to drain.  

I don't see much value in a control to enable or disable pipelining at the work-item level because that's nearly always beneficial.  This invocation pipelining is more situational though as it only provides value if the user intends to launch the same kernel multiple times.